### PR TITLE
fix(evaluation): make spectral_matrix_sign scale-free via power-of-two rescaling

### LINKED ISSUE #2379: bug(optimizer-geometry): the 1e-12 Frobenius floor launders small matrices into a numerically zero matrix sign
## Summary

`spectral_matrix_sign` divides by a fixed `1e-12` magnitude floor instead of the matrix's own Frobenius norm, so any matrix whose norm falls below that floor is **laundered into a numerically zero "valid" sign** rather than orthogonalized. The function's docstring promises the opposite:

> Frobenius normalization places every singular value in its convergence interval and preserves an exact zero for a zero matrix.

`alberta_framework/evaluation/optimizer_geometry.py:174-176`:

```python
norm = jnp.linalg.norm(value)
valid = jnp.all(jnp.isfinite(value)) & jnp.isfinite(norm)
x = value / jnp.maximum(norm, jnp.asarray(1e-12, dtype=value.dtype))
```

When `norm < 1e-12` the divisor is the floor, not the norm, so `‖x‖_F = ‖M‖_F / 1e-12 ≪ 1`. The NS5 recurrence `f(X) = 3/2 X − 1/2 XX^T X` is then applied to singular values far below its convergence interval: each step multiplies them by ≈1.5, so five steps return ≈`7.59 · σ / 1e-12` — a matrix that is numerically zero — while the transaction still reports `valid=True`.

In float32 the floor is reached earlier than the exponent range suggests, because the sum of squares underflows: for entries near `1e-20`, `jnp.linalg.norm` returns exactly `0.0` for a plainly nonzero matrix.

## Reproduction

```python
import jax
import jax.numpy as jnp
from alberta_framework.evaluation.optimizer_geometry import spectral_matrix_sign

m = jnp.array([[2.0, 1.0], [0.5, -1.0]], dtype=jnp.float32)
for scale in (1.0, 1e-13, 1e-15, 1e-20, 1e-25):
    tiny = m * jnp.asarray(scale, dtype=jnp.float32)
    print(scale, float(jnp.linalg.norm(tiny)), float(jnp.linalg.norm(spectral_matrix_sign(tiny))))
```

```
scale     norm(input)      norm(returned sign)   expected
1e+00     2.500000e+00     1.414202e+00          1.414202e+00
1e-13     2.500000e-13     1.201002e+00          1.414202e+00
1e-15     2.500000e-15     1.898316e-02          1.414202e+00
1e-20     0.000000e+00     1.898437e-07          1.414202e+00
1e-25     0.000000e+00     1.898438e-12          1.414202e+00
```

Every row is the same matrix up to a positive scalar, so every row must return the same sign. `valid` is `True` in all five.

Singular values of the returned "sign" (computed in float64):

| input scale | singular values | orthogonal? |
| --- | --- | --- |
| `1.0` | `[0.99999999, 0.99998346]` | yes |
| `1e-13` | `[0.96671583, 0.71264756]` | no |
| `1e-25` | `[1.69801409e-12, 8.49007043e-13]` | no — numerically zero |

The reverse-mode derivative is also lost: `jax.jacrev(spectral_matrix_sign)` is non-finite at `1e-25` scale and finite at unit scale.

## Consumer impact

`muon_ogd_dual_update_transaction` uses this sign as the update direction (`optimizer_geometry.py:260` and `:274`). On the documented empty-constraint path — pinned by `test_muon_ogd_empty_constraints_reduces_exactly_to_ns5` as reducing exactly to NS5 — the update vanishes with gradient scale instead of staying scale-free:

| gradient scale | `‖update‖_F` |
| --- | --- |
| `1.0` | `1.414202e+00` |
| `1e-13` | `1.201002e+00` |
| `1e-25` | `1.898438e-12` |

A scale-free orthogonalized direction is the entire point of the Muon-OGD step, so a small-magnitude momentum matrix silently produces no step rather than a unit-scale one.

The frozen streaming benchmark does not expose this: `_frozen_stream()` matrices have Frobenius norms between `1.15` and `2.59`, all far above the floor, which is why every published arm metric is unaffected.

## Why the floor is the wrong instrument

The existing `test_geometry_float32_overflow_is_invalid_not_laundered` pins the policy for the other end of the exponent range: when the input's own arithmetic overflows, the transaction must report `valid=False` rather than launder a plausible-looking result. The small end does the opposite — it returns a fabricated near-zero matrix with `valid=True`. A magnitude floor cannot distinguish "genuinely zero" from "small but well-conditioned"; only the norm can, and the norm is unusable in the input dtype at that magnitude.

## Proposed fix

Rescale by an exact power of two before normalizing, which is the pattern already used for the same hazard in `alberta_framework/utils/nexting.py`, `alberta_framework/core/oak.py`, `alberta_framework/core/state_builder.py`, and `alberta_framework/benchmarks/ipmnist_screening.py`:

```python
_, exponent = jnp.frexp(jnp.max(jnp.abs(value)))
rescaled = jnp.ldexp(value, -exponent)
rescaled_norm = jnp.linalg.norm(rescaled)
positive = rescaled_norm > 0
x = jnp.where(
    positive,
    rescaled / jnp.where(positive, rescaled_norm, jnp.ones_like(rescaled_norm)),
    jnp.zeros_like(rescaled),
)
```

The largest entry lands in `[0.5, 1)`, so the sum of squares can neither overflow nor underflow, and an exact zero is preserved by the zero test rather than by a magnitude floor. Because power-of-two scaling is exact, the quotient is bit-identical to `value / norm` for every input whose norm was already representable — verified bit-identical on 200 random float32 matrices spanning `2**-30` to `2**30`. `norm` stays in place as the overflow signal so the pinned invalid-on-overflow behaviour is unchanged.

## Environment

JAX 0.11.0, jaxlib 0.11.0, numpy 2.5.1, CPython 3.13.14, Windows 11 AMD64, `jax_enable_x64` False (default).

### PR DESCRIPTION
**Pull request by Kayvan-Zahiri**
Fixes #2379

`spectral_matrix_sign` previously normalized inputs using a fixed `1e-12` Frobenius magnitude floor: `x = value / max(norm, 1e-12)`. When `norm < 1e-12`, the NS5 cubic recurrence $f(X) = \frac{3}{2}X - \frac{1}{2}X X^T X$ was applied to singular values far below its convergence interval, laundering small matrices into numerically zero matrix signs while returning `valid=True`.

### Changes
1. Replaced the `1e-12` magnitude floor with exact power-of-two exponent rescaling:
   ```python
   max_abs = jnp.max(jnp.abs(value))
   _, exponent = jnp.frexp(max_abs)
   rescaled = jnp.ldexp(value, -exponent)
   rescaled_norm = jnp.linalg.norm(rescaled)
   ```
   This centers input entries in $[0.5, 1.0)$, placing all singular values within the NS5 convergence basin and making the orthogonalization scale-free from $1.0$ down to $10^{-25}$.
2. Added unit test `test_spectral_matrix_sign_scale_freedom` in `tests/test_optimizer_geometry.py` verifying that small matrices ($10^{-13}$ to $10^{-25}$) return the exact orthogonal polar factor matching unit scale.

### Verification
- `pytest tests/test_optimizer_geometry.py`: 29/29 tests passed.
- `ruff check`: All checks passed.
- `mypy`: Clean.

### CODE REVIEW

**Review by saltwire-dev — CHANGES_REQUESTED** [review]
## Request changes at exact head `01424aec81e565c27b8d1ad93e90d58a1265c3de`

Reviewed at exact head `01424aec81e565c27b8d1ad93e90d58a1265c3de` against production `alberta_framework/evaluation/optimizer_geometry.py` (not just the new test). Conversation sidebar is No reviews; the github-actions “produced no report” note is a comment, not a review. JAX 0.11.1 CPU, `jax_enable_x64=False`, `jax.jit` of this file.

Power-of-two rescaling is the right repair for #2379’s *all-normal tiny* rows — on this CPU, float32 `‖M‖_F` really is `0` at scale `1e-20`/`1e-25`, and this helper really does flip those rows from a numerically zero “valid” sign (`‖S‖_F ~ 0` under `jax.jit` on main) to `‖S‖_F = 1.4142019` with singular values `[0.99999999, 0.99998346]`, bit-identical to the unit-scale NS5 sign. The hole is that the rescale is **unconditional**, so JAX `ldexp`’s documented “subnormal operand returned unmodified” behavior rotates mixed normal/subnormal matrices into a **rank-1** iterate. NS5 then emits a plausible unit-scale `[[1,0],[0,0]]` with `valid=True` — **worse laundering than main’s zeros** — while the new test never drives a subnormal, the public wrapper, or the Muon-OGD consumer.

**Blocking: unconditional `frexp`/`ldexp` rotates mixed-subnormal matrices and launders `valid=True`**

Issue #2379’s proposed patch is exactly what landed:

```python
# optimizer_geometry.py:169-180
norm = jnp.linalg.norm(value)
max_abs = jnp.max(jnp.abs(value))
_, exponent = jnp.frexp(max_abs)
rescaled = jnp.ldexp(value, -exponent)
rescaled_norm = jnp.linalg.norm(rescaled)
positive = rescaled_norm > 0
valid = jnp.all(jnp.isfinite(value)) & jnp.isfinite(norm) & jnp.isfinite(rescaled_norm)
```

No gate on “raw Frobenius already live”. No subnormal-exponent withhold. `valid` does not include `positive`, so a rotated (or fully flushed) rescale is still reported valid.

Independent run, issue matrix `M=[[2,1],[0.5,-1]]` (float32):

| input | raw `‖M‖_F` f32 | this head `valid` | `‖S‖_F` | vs main (`1e-12` floor) |
|---|---|---|---|---|
| `1e-20` / `1e-25` / `1e-30` | 0 | True | 1.4142019 | **repaired** |
| **`1e-38` mixed** | **0** | **True** | **1.0** (`[[1,0],[0,0]]`, err **0.894**, svs `[1,0]`) | **worse than main’s zeros** |
| `2.938736e-39` all-subnormal | 0 | **True** | **0** | unchanged zeros, still `valid=True` |
| `1e20` / `1e30` / `1.7e38` | inf | False | 0 | overflow still refused; **untested** |

Mixed `1e-38` internals (numpy-first so JAX does not FTZ the scale):

```
input  bits : ['0x00d9c7dc', '0x006ce3ee', '0x003671f7', '0x806ce3ee']
frexp(max_abs) exponent = -125
rescaled bits : ['0x3f59c7dc', '0x006ce3ee', '0x003671f7', '0x806ce3ee']
unchanged by rescale: [False, True, True, True]
safe     = [[1.0, 0.0], [0.0, -0.0]]   # NS5 sign of e00, not of M
valid    = True
```

All-subnormal: `frexp` reports exponent 0, `ldexp` is a no-op, `rescaled_norm == 0`, `valid` stays True. Public `spectral_matrix_sign` returns zeros with no raise. Empty-constraint Muon inherits both: mixed yields `‖update‖_F=1.0` `valid=True` (wrong direction, looks live); all-subnormal yields silent no-step.

**Required fix:** (1) rescale **only** when raw Frobenius is exactly `0` or non-finite; (2) withhold rescale from any matrix with a subnormal exponent field — mixed must not become rank-1 `[[1,0],[0,0]]`, all-subnormal must not report `valid=True` zeros; (3) keep overflow `valid=False`; (4) share one helper with #2430 / #2393 if both land — do not open-code a third un-gated `frexp`/`ldexp`.

**Blocking: tests pin the naive closed form, not the rows that falsify it**

New `test_spectral_matrix_sign_scale_freedom` loops `(1e-13, 1e-15, 1e-20, 1e-25)` only — all-normal tiny scales. Missing: mixed `1e-38` (this head **regresses**), all-subnormal, overflow, float16 `2**-20`, public `spectral_matrix_sign`, empty-constraint Muon, bit-identity of unit-scale vs `main`.

**Required fix:** parametrize the helper *and* `spectral_matrix_sign` *and* the empty-constraint Muon path with the issue table plus mixed/all-subnormal withhold rows. Assert `valid=False` (or zeros with `valid=False`) for mixed/all-subnormal until a subnormal-safe rescale exists.

**Verified**

- Human PR; Conversation sidebar is literally No reviews.
- Head SHA `01424aec81e565c27b8d1ad93e90d58a1265c3de` matches the `.patch` From-line.
- Algebra of `sign(cM)=sign(M)` is correct **when `ldexp` moves every finite entry**.
- Unit-scale is bit-identical to main. Float16 `2**-20` is repaired here (`‖S‖_F≈1.415`) but untested.
- Later same-theme clones still No reviews; they do not scoop this head.

I would approve a follow-up that gates rescale on dead/non-finite Frobenius, withholds subnormal exponent fields, and drives mixed `1e-38` plus all-subnormal fail-closed. Until then this is the naive un-gated sketch that turns mixed-subnormal matrices into a plausible rank-1 “sign”.

**Comment by lalalune** [comment]
Superseded by merged #2380. Reviewed this matrix-sign rescaling diff: exact power-of-two normalization now replaces the fixed 1e-12 floor. The selected implementation has 17 passing regressions covering scale invariance, JIT, gradients, float16/bfloat16, zero input, and the Muon-OGD consumer, with required CI green. Keeping another equivalent normalization branch would add no useful behavior.
